src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -161,7 +161,10 @@ public synchronized void enableAutorelease(FinalizationService finalizationServi
             return;
         }
 
-        finalizationService.addFinalizer(this, Pointer.class, () -> free());
+        // We must be careful here that the finalizer does not capture the Pointer itself that we'd
+        // like to finalize.
+        long pointer = address;
+        finalizationService.addFinalizer(this, Pointer.class, () -> UNSAFE.freeMemory(pointer));
```

Comment:
Subtle!

Reply:
Maybe it would be safer to use the pattern like in Ruby with an extra static method returning a Runnable, to which we pass the address explicitly. Do the semantics in Java guarantee to not capture `this` if it's not used?